lib/apimlStorage.ts

```diff
@@ -0,0 +1,382 @@
+    typeof response.json.messages === 'object' &&
+    typeof response.json.messages[0] === 'object' &&
+    typeof response.json.messages[0].messageKey == 'string'
+  ) {
+    return response.json.messages[0].messageKey;
+  }
+  return undefined;
+}
+
+async function apimlDoRequest(req: ApimlRequest): Promise<ApimlResponse> {
+  if (!apimlClient) {
+    throw new ApimlStorageError('APIML_STORAGE_NOT_CONFIGURED');
+  }
+  try {
+    const response = await apimlClient.request({
+      method: req.method,
+      url: req.path,
+      data: req.body,
+      headers: req.headers,
+    });
+    const apimlResponse: ApimlResponse = {
+      headers: response.headers,
+      statusCode: response.status,
+      json: response.data
+    };
+    return apimlResponse;
+  } catch (e) {
+    if (e.response) {
+      const response = e.response;
+      const apimlResponse: ApimlResponse = {
+        headers: response.headers,
+        statusCode: response.status,
+        json: response.data
+      };
+      const err = checkHttpResponse(apimlResponse);
+      if (err) {
+        throw err;
+      }
+      return apimlResponse;
+    } else if (e.request) {
+      throw new ApimlStorageError('APIML_STORAGE_CONNECTION_ERROR', e);
+    } else {
+      throw new ApimlStorageError('APIML_STORAGE_UNKNOWN_ERROR', e);
+    }
+  }
+}
+
+function checkHttpResponse(response: ApimlResponse): ApimlStorageError | undefined {
+  switch (response.statusCode) {
+    case 404: /* HTTP_STATUS_NOT_FOUND */
+      return new ApimlStorageError('APIML_STORAGE_KEY_NOT_FOUND', undefined, response);
+    case 403: /* HTTP_STATUS_FORBIDDEN */
+      return new ApimlStorageError('APIML_STORAGE_UNKNOWN_CLIENT_CERT', undefined, response);
+    case 401: /* HTTP_STATUS_UNAUTHORIZED */
+      return new ApimlStorageError('APIML_STORAGE_UNAUTHORIZED', undefined, response);
+    case 400: /* HTTP_STATUS_BAD_REQUEST */
+      const errorKey = apimlResponseGetMessageKey(response);
+      if (errorKey === 'org.zowe.apiml.cache.invalidPayload') {
+        return new ApimlStorageError('APIML_STORAGE_INVALID_PAYLOAD', undefined, response);
+      } else if (errorKey === 'org.zowe.apiml.security.query.tokenNotProvided') {
+        return new ApimlStorageError('APIML_STORAGE_NO_AUTH_PROVIDED', undefined, response);
+      } else {
+        return new ApimlStorageError('APIML_STORAGE_UNKNOWN_ERROR', undefined, response);
+      }
+    case 200: /* HTTP_STATUS_OK */
+      return;
+    case 204: /* HTTP_STATUS_NO_CONTENT */
+      return;
+    case 201: /* HTTP_STATUS_CREATED */
+      return;
+    default:
+      return new ApimlStorageError('APIML_STORAGE_UNKNOWN_ERROR', undefined, response);
+  }
+}
+class ApimlStorage {
+  private keyPrefix: string;
+
+  constructor(
+    private pluginId: string
+  ) {
+    this.keyPrefix = this.makeKeyPrefix(this.pluginId);
+  }
+
+  private makeKeyPrefix(pluginId: string): string {
+    return pluginId.replace(/\./g, '');
```

Comment:
this will cause collisions when products have poorly organized names.
needs a code comment linking to https://github.com/zowe/api-layer/issues/1300

Reply:
I added a comment